cmd/cli/bulk.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/multiplay/go-svrquery/lib/svrquery"
+	"github.com/multiplay/go-svrquery/lib/svrquery/protocol"
+)
+
+const (
+	numWorkers = 100
+)
+
+var (
+	errNoItem       = errors.New("no item")
+	errEntryInvalid = errors.New("invalid entry")
+)
+
+// BulkResponseItem contains the information about the query being performed
+// against a single server.
+type BulkResponseItem struct {
+	Address    string                      `json:"address"`
+	ServerInfo *BulkResponseServerInfoItem `json:"serverInfo,omitempty"`
+	Error      string                      `json:"error,omitempty"`
+}
+
+// BulkResponseServerInfoItem container the server information.
+type BulkResponseServerInfoItem struct {
+	CurrentPlayers int64  `json:"currentPlayers"`
+	MaxPlayers     int64  `json:"maxPlayers"`
+	Map            string `json:"map"`
+}
+
+// BulkResponseItemWork is an item returned by an worker containing the data item
+// plus any terminal error it encountered.
+type BulkResponseItemWork struct {
+	Item *BulkResponseItem
+	Err  error
+}
+
+// queryBulk queries a bulk set of servers using a query file.
+func queryBulk(file string) error {
+	f, err := os.Open(file)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	// Make a jobs channel and a number of workers to processes
+	// work off of the channel.
+	jobChan := make(chan string)
+	resultsChan := make(chan BulkResponseItemWork)
+	for w := 1; w <= numWorkers; w++ {
+		go worker(jobChan, resultsChan)
+	}
+
+	items := make([]BulkResponseItem, 0, 1000)
+
+	// Queue work onto the channel
+	scanner := bufio.NewScanner(f)
+	jobCount := 0
+	for scanner.Scan() {
+		jobCount++
+		jobChan <- scanner.Text()
+	}
+
```

Comment:
Missing close of jobChan?

Reply:
Added a wg plus proper tidyup here now

Reply:
If you close the `jobChan` here instead of below, you don't need the `WaitGroup` for the flow below.

However if you keep it you could use the following to allow you to range over the results channel instead.

```go
go func() {
   defer close(resultsChan)
   wg.Wait()
}()
```

Pick your poison :)